### PR TITLE
DX: Increase PHPStan level to 6

### PR DESCRIPTION
**Pull request by julienfalque**

### CODE REVIEW

File: phpstan.neon [hunk 1]
```diff
@@ -14,5 +14,17 @@ parameters:
...
+        - '#^Method PhpCsFixer\\Tests\\.+::provide.+Cases\(\) return type has no value type specified in iterable type array\.$#'
+        - '#^Method PhpCsFixer\\Tests\\.+::provide.+Cases\(\) return type has no value type specified in iterable type iterable\.$#'
+
+        # baseline
+        - '#^Method .+ has parameter .+ with no value type specified in iterable type array\.$#'
+        - '#^Method .+ has parameter .+ with no value type specified in iterable type iterable\.$#'
+        - '#^Method .+ return type has no value type specified in iterable type array\.$#'
+        - '#^Method .+ return type has no value type specified in iterable type iterable\.$#'
+        - '#^Property .+ has no type specified\.$#'
+        - '#^Property .+ type has no value type specified in iterable type array\.$#'
```
Comment (jrmajor) [inline phpstan.neon:28]: Please consider generating `phpstan-baseline.neon` instead (or adding `count`). It may be huge, but it will ensure that no new errors of these types will be introduced.

If you decide to stick with the current approach, you can use `checkMissingIterableValueType: false` parameter.

File: src/Runner/FileFilterIterator.php [hunk 2]
```diff
@@ -54,15 +57,6 @@ public function __construct(
     public function accept(): bool
     {
         $file = $this->current();
-        if (!$file instanceof \SplFileInfo) {
```
Comment (keradus) [inline src/Runner/FileFilterIterator.php:57]: tbh I'm not sure. one can ignore param annotation and pass whatever iterator they want.
until PHP starts support templated types on language level, i would keep the check here.

Comment (jrmajor) [reply src/Runner/FileFilterIterator.php:57]: You can't check the content of every iterable... I'd say until PHP starts supporting templated types on the language level, we need to rely on static analysis :)

File: src/Fixer/Operator/TernaryToElvisOperatorFixer.php [hunk 3]
```diff
@@ -29,7 +29,7 @@ final class TernaryToElvisOperatorFixer extends AbstractFixer
      *
      * Ordered by most common types first.
      *
-     * @var array
+     * @var list<array{0: int}|string>
```
[the remark this answers is not in the document]
Comment (julienfalque) [reply src/Fixer/Operator/TernaryToElvisOperatorFixer.php]: Done, thx!

File: phpstan.neon (hunk 1 above)
Comment (julienfalque) [reply phpstan.neon]: Letting PHPStan generate the baseline was my first approach when I previously tried to increase PHPStan's level. I think baseline feature is great but it tends to create a burden on contributors that often have to update it, so not very suitable for us here.

I added counts, thanks for the suggestion.

File: src/Runner/FileFilterIterator.php (hunk 2 above)
Comment (julienfalque) [reply src/Runner/FileFilterIterator.php:57]: I would agree and keep the check for public APIs where we have no guarantee downstream users will use static analysis, but this class is internal so static analysis is sufficient.

Comment (jrmajor) [reply src/Runner/FileFilterIterator.php:57]: Actually, it receives a finder from the user config with no type check in between, and `ConfigInterface::setFinder()` does not specify iterable type, so downstream user can pass any iterable easily and wouldn't even break the API.

Comment (julienfalque) [reply src/Runner/FileFilterIterator.php]: Right, good catch!

Comment (VincentLanglet) [reply src/Runner/FileFilterIterator.php]: You can use https://phpstan.org/config-reference#treatphpdoctypesascertain in the phpstan config, if you want to still be able to do check like this

File: src/Runner/FileLintingIterator.php [hunk 4]
```diff
@@ -21,6 +21,8 @@
  * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
  *
  * @internal
+ *
+ * @extends \IteratorIterator<mixed, \SplFileInfo, \Traversable<\SplFileInfo>>
```
Comment (keradus) [inline src/Runner/FileLintingIterator.php:25]: what is the first type here, for? maybe we can avoid `mixed` here?

File: tests/Runner/FileLintingIteratorTest.php [hunk 5]
```diff
@@ -37,7 +37,6 @@ public function testFileLintingIteratorEmpty(): void
             $fileLintingIteratorProphecy->reveal()
         );
 
-        static::assertNull($fileLintingIterator->current());
```
Comment (keradus) [inline tests/Runner/FileLintingIteratorTest.php:40]: I'm wondering, why this change?

File: src/Runner/FileLintingIterator.php (hunk 4 above)
Comment (julienfalque) [reply src/Runner/FileLintingIterator.php:25]: It's mandatory. I used `mixed` because we don't actually care about the key type, but I might change it if you want.

File: tests/Runner/FileLintingIteratorTest.php (hunk 5 above)
Comment (julienfalque) [reply tests/Runner/FileLintingIteratorTest.php:40]: PHPStan reported that the assertion cannot pass and I removed it too fast. I restored it, I think the error is actually a false positive:
```
------ -------------------------------------------------------------------------------------------------------------- 
  Line   tests/Runner/FileLintingIteratorTest.php                                                                      
 ------ -------------------------------------------------------------------------------------------------------------- 
  40     Call to static method PHPUnit\Framework\Assert::assertNull() with SplFileInfo will always evaluate to false.  
 ------ --------------------------------------------------------------------------------------------------------------
```

File: phpstan.neon (hunk 1 above)
Comment (jrmajor) [reply phpstan.neon]: I believe that baseline would be easier to deal with for contributors, as they get a suitable error message asking them to add type or remove something from baseline. With count, I think they will get something like “expected to get this error x times, but it was reported y times”, with no way to know where the change occurred.

Comment (julienfalque) [reply phpstan.neon]: I propose to see how it goes and if it still makes contributions harder, we'll try something else.
